darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -0,0 +1,124 @@
+        amplify=False):
+    """
+    Generates a heatmap plot from a report with DXT traces.
+
+    Args:
+        report (darshan.DarshanReport): report to generate plot from
+        xbins (int): number of bins on the x axis
+        ybins (int): number of bins on the y axis
+        group_by (str): attribute to group by (e.g., rank, hostname)
+        mods (list): modules to include in heatmap (e.g., ['DXT_POSIX', 'DXT_MPIIO'])
+        ops (list): operations to consider (e.g., ['read', 'write']
+        display_values (bool): show values per heatmap field
+        cmap: overwrite colormap (see matplotlib colormaps)
+        figsize: change figure size (see matplotlib figsize)
+
+        amplify (int): paint neighbouring cells e.g., when working with many ranks
+    """
+
+    runtime = report.end_time - report.start_time       # timedelta
+    runtime = runtime.seconds + runtime.microseconds    # to float/int
+    if runtime == 0:
+        runtime = 1
+
+        
+    if mods is None:
+        mods = ['DXT_POSIX', 'DXT_MPIIO']
+    
+    if ops is None:
+        ops = ['read', 'write']
+    
+    
+    def get_hostname_mapper(report):
+        """ Determine which ranks map to which hostname. """
+        hostnames = {}
+        i = 0
+        for mod in mods:
+            for rec in report.records[mod]:
+                hostname = rec['hostname']
+                if hostname not in hostnames:
+                    hostnames[hostname] = i
+                    i += 1
+        return hostnames
+   
+
+    hostnames = get_hostname_mapper(report)
+    ranks = report.metadata['job']['nprocs']
+    
+    # set ybins, yticks and ylabel to match choice of group_by 
+    if group_by in ['hostname', 'node']:
+        ybins = len(hostnames)
+        yticks = list(hostnames.keys())
+        ylabel = "Hostname"
+    else:
+        if ybins is None:
+            ybins = ranks 
+            ylabel = "Rank"
+            #yticks = [str(x) for x in range(ybins)] 
+        else:
+            ylabel = f"Binned Ranks (binsize={int(ranks/ybins)} ranks)"
+
+    
+    # heatmap to be populated with event counts
+    events = np.zeros((ybins, xbins))    
+    
+    for mod in mods:
+        for op in ops:
+            for rec in report.records[mod]:
+                
+                if group_by in ['hostname', 'node']:
+                    ybin = hostnames[rec['hostname']]
+                else:
+                    ybin = int((rec['rank'] /ranks) * ybins)
+                    
+                    
+                for event in rec[f'{op}_segments']:
+                    xbin = int((event['start_time'] / runtime) * xbins)
+                    events[ybin][xbin] += 1
+                    if amplify:
+                        rng = amplify
+                        for i in range(rng):
+                            sur = int(rng/2) + i
+                            if sur != 0 and (ybin - sur) > 0:
+                                events[ybin - sur][xbin] += 1
```

Comment:
I'd probably try to nudge us in the direction of moving processing/number crunching code to functions under test, along with some i.e., DXT files we can use to guard against regressions, etc.

May also be helpful to have a function that describes what the looping is doing (input/expected output, etc.).

Reply:
I agree very much. The pre-processing for plots should be as much as possible decoupled to allow custom analysis on top of the aggregations. 

For this particular snippet, it's really is just a quick (well actually slow) workaround to amplify visibility, e.g., when dealing with high rank numbers. There is certainly better ways to handle these.   Here I would not break it aways from the plot to much because this is purely for visuals, because the matrix is not holding true values for neighboring ranks anymore after applying it.  Much room for different optimizations for sure.  Allowing to switch log scaling for the intensities/transmitted bytes might be one.